doc/Makefile

```
@@ -1,82 +1,135 @@
 
 byte:
-	export HEVEADIR=.. ;\
-	$(MAKE) $(MFLAGS) HEVEAOPTS="$(HEVEAOPTS)" HACHAOPTS="$(HACHAOPTS)" HEVEA="../hevea.byte" HACHA=../hacha.byte IMAGEN=../imagen all
+	HEVEADIR=.. $(MAKE) $(MFLAGS) HEVEAOPTS="$(HEVEAOPTS)" HACHAOPTS="$(HACHAOPTS)" HEVEA=../hevea.byte HACHA=../hacha.byte IMAGEN=../imagen all
 
 tmp.tex: text.tex
-	$(GPIC) -t < text.tex > tmp.tex
+	$(GPIC) -t < $< > $@
+
+MANUAL_DEPENDENCIES = manual.tex macros.tex tmp.tex version.tex # "manual.tex" MUST be first!
 
-doc/manual.html: doc manual.hva  macros.tex manual.tex version.tex tmp.tex
-	$(HEVEA) $(HEVEAOPTS) -o doc/manual.html manual.hva manual.tex
+doc/manual.html: doc manual.hva $(MANUAL_DEPENDENCIES)
+	$(HEVEA) $(HEVEAOPTS) -o $@ manual.hva manual.tex
 
-manual.dvi: manual.tex macros.tex version.tex tmp.tex
-	$(LATEX) < /dev/null manual.tex &&\
-	$(LATEX) < /dev/null manual.tex &&\
-	makeindex manual.idx &&\
-	$(LATEX) < /dev/null manual.tex
+doc/index.html: doc/manual.html
+	$(HACHA) $(HACHAOPTS) -o $@ $<
+
+LATEX_LABEL_WARNING = 'LaTeX Warning: Label(s) may have changed'
+
+manual.dvi: $(MANUAL_DEPENDENCIES)
+	$(LATEX) $(LATEXOPTS) $<
+	$(LATEX) $(LATEXOPTS) $< > /dev/null 2>&1
+	@ITER=2; MAXITER=10;  \
+        while test -e $*.log && grep -q $(LATEX_LABEL_WARNING) $*.log; do  \
+            test $$ITER -le $$MAXITER || { printf 'Excessive number of iterations (%d) to reach fixpoint.  Aborted!\n' "$$MAXITER" 1>&2; exit 1; };  \
+            printf 'Run LaTeX again.\n'  \
+            $(LATEX) $(LATEXOPTS) $< > /dev/null 2>&1;  \
+            : $$((ITER += 1));  \
+	done;  \
+        printf 'Fixpoint reached after %d iteration(s).\n' "$$ITER"
 
 manual.ps: manual.dvi
-	dvips -o manual.ps manual.dvi
+	$(DVIPS) $(DVIPSOPTS) -o $@ $<
 
-manual.pdf: manual.ps
-	ps2pdf manual.ps
+manual.pdf: $(MANUAL_DEPENDENCIES)
+	$(PDFLATEX) $(PDFLATEXOPTS) $<
+	$(PDFLATEX) $(PDFLATEXOPTS) $< > /dev/null 2>&1
+	@ITER=2; MAXITER=10;  \
+        while test -e $*.log && grep -q $(LATEX_LABEL_WARNING) $*.log; do  \
+            test $$ITER -le $$MAXITER || { printf 'Excessive number of iterations (%d) to reach fixpoint.  Aborted!\n' "$$MAXITER" 1>&2; exit 1; };  \
+            printf 'Run PDFLaTeX again.\n'  \
+            $(PDFLATEX) $(PDFLATEXOPTS) $< > /dev/null 2>&1;  \
+            : $$((ITER += 1));  \
+	done;  \
+        printf 'Fixpoint reached after %d iteration(s).\n' "$$ITER"
 
-doc/index.html: doc/manual.html
-	$(HACHA) $(HACHAOPTS) -o doc/index.html doc/manual.html
+manual.info: $(MANUAL_DEPENDENCIES)
+	$(HEVEA) $(HEVEAOPTS) -info -o $@ $<
 
-info: manual.inf tmp.tex
-	$(HEVEA) $(HEVEAOPTS) -info manual.inf manual.tex -o manual.info
+manual.txt: $(MANUAL_DEPENDENCIES)
+	$(HEVEA) $(HEVEAOPTS) -text -o $@ $<
 
-thai: doc doc/thaihevea.html
+.PHONY: thai
+thai: doc/thaihevea.html
 
-doc/thaihevea.html: thai/thaihevea.ttex
-	export TEXINPUTS="thai:" ; $(HEVEA)  $(HEVEAOPTS) -o $@ $<
+doc/thaihevea.html: doc thai/thaihevea.ttex
+	TEXINPUTS="thai:" $(HEVEA) $(HEVEAOPTS) -o $@ $<
```

For build to work smoothly on my system, I had to peform the following change.
```suggestion
	TEXINPUTS="thai:" $(HEVEA) $(HEVEAOPTS) -o $@ thai/thaihevea.ttex
```